### DISCUSSION TITLE
How to add calendar with one ics file per event?

### DISCUSSION THREAD

**Discussion by kflak** [kiwanami/emacs-calfw#166] [Q&A] [answered]
Hi,

I'm dipping my toes into the emacs waters after years of vimming :-) and trying now to get my calendar setup working. I use `vdirsyncer` to sync my calendars from a nextcloud instance, and I am getting one ics file per event in that calendar. It's not clear to me how to import this into cfw.... I tried using a wildcard, but that didn't work out too well...:
```
    (cfw:ical-create-source "Calendar" "~/path/to/calendar/*.ics" "Gray")  ; 
```
Any help much appreciated!

**Comment by hperrey** [comment]
I unfortunately do not know how to achieve this with `calfw`. Just in case you have not already solved it or for future visitors I just want to bring up an alternative: this is the exact setup for which I wrote `khalel` which allows me to interact from Org mode with calendars synced via `vdirsyncer`: https://gitlab.com/hperrey/khalel
If you already have `vdirsyncer` set up, you only need to install the Python package `khal` and you have all the prerequisites in place :)

**Comment by kflak** [comment]
Hi there, thanks for the tip! I use khal as well, and my emacs experiment ended in ignomy and I slinked back to my beloved neovim :-)

**Comment by kflak** [comment]
Trying again to get this setup to work, so that I would have a fully functional offline experience... I noticed it is closed, so I was curious if someone has an example configuration for making this work?

**Comment by haji-ali** [comment]
Try this (untested)

```lisp
(defun calfw-ical-create-source (name urls color)
  "Create a calfw source for an ical calendar at URL(s).

NAME is the name of the calendar, COLOR is the color to use for
all URLs. Returns a calfw source."
  (let ((urls (ensure-list urls)))
    (make-calfw-source
     :name (concat "iCal:" name)
     :color color
     :update (lambda () (mapc
                         (lambda (url) (calfw-ical-data-cache-clear url))
                         urls))
     :data (lambda (begin end)
             (mapcan
              (lambda (url) (calfw-ical-to-calendar url begin end))
              urls)))))
```

and use like this

```
    (calfw-ical-create-source "Calendar" (directory-files "~/path/to/calendar/" t "\\.ics$") "Gray")
```

If it works for you, I might add it to calfw (though it would require some thorough testing).

**Comment by kflak** [comment]
Hmmm... trying this:

```lisp
(defun calfw-ical-create-source (name urls color)
  "Create a calfw source for an ical calendar at URL(s).

NAME is the name of the calendar, COLOR is the color to use for
all URLs. Returns a calfw source."
  (let ((urls (ensure-list urls)))
    (make-calfw-source
     :name (concat "iCal:" name)
     :color color
     :update (lambda () (mapc
                         (lambda (url) (calfw-ical-data-cache-clear url))
                         urls))
     :data (lambda (begin end)
             (mapcan
              (lambda (url) (calfw-ical-to-calendar url begin end))
              urls)))))

(defun my-open-calendar ()
  (interactive)
  (calfw-open-calendar-buffer
   :contents-sources
   (list
    (calfw-ical-create-source "Calendar" (directory-files "~/.calendars/webo/personal/" t "\\.ics$") "Gray"))))
```

Gives me the error:
```
Symbol’s value as variable is void: urls
```

I have verified that the directory is correct and full of .ics files.

**Reply by haji-ali** [reply]
Do you have lexical-binding turned on?
A stack trace could be helpful otherwise

**Comment by kflak** [comment]
Lexical-binding was off. Setting it to t makes it work! However, it takes quite a loooong time to load the calendar events. Is there a way to cache the result, so that subsequents calls to the function would be faster?

**Comment by kflak** [comment, marked as answer]
Just realized that a much less expensive option is to merge all the ics files into one. I found this script that I can modify to suit my needs: https://gist.github.com/HappyStinson/e8e39bfa8405993b2251eaa73e47af70. Adding this to the systemd service that I am using to pull my calendar at regular intervals is trivial, which obviates the need to open thousands of ics files in emacs...
